### PR TITLE
Old ore processor cleanup

### LINKED ISSUE #23093: Old ore smelter doesnt drop ore if destroyed
image

SCREAMS INTERNALLY

I didnt want to smelt it all but i couldnt eject it

### LINKED ISSUE #25166: Ore refiner can't be used to refine titanium and plastitanium
While ORM is now the main way of refining ores, the old ore refiners still exist you know?

Ore refiners at the mining outpost and labor camp ignore titanium ores.

### PR DESCRIPTION
**Pull request by Profakos**
The old ore processor can now smelt titanium, and also, bluespace crystal. It also now Initializes and has a proximity monitor. 

Also updated  /obj/machinery/mineral/proc/unload_mineral to use forceMove.

fixes #25166
fixes #23093
I tried to make this code dynamic, like the one in the stacking machine, but I couldn't figure out how to do that and still be able to handle making plastitanium and plasteel. 
e:
here I go, into the depths, refactoring this entire hell machine

e2: all done! If anything needs to be tweaked, I'll  back on the 27th to do it.

### CODE REVIEW

File: code/modules/mining/machine_processing.dm [hunk 1]
```diff
@@ -215,98 +229,157 @@
...
 
+/obj/machinery/mineral/processing_unit/Initialize()
+	. = ..()
+	proximity_monitor = new(src, 1)
+
+/obj/machinery/mineral/processing_unit/HasProximity(atom/movable/AM)
+	if(istype(AM, /obj/item/weapon/ore) && AM.loc == get_step(src, input_dir))
+		process_ore(AM)
+
+/obj/machinery/mineral/processing_unit/proc/process_ore(O)
```
Comment (xxalpha) [inline code/modules/mining/machine_processing.dm]: 👨‍💻 🔫 😡 💢

Comment (Profakos) [inline code/modules/mining/machine_processing.dm]: Oh no, I forgot to type it, fixing this asap.
Or maybe move this proc, along with the stacker's, up a level to the parent.

Comment (xxalpha) [inline code/modules/mining/machine_processing.dm]: That is not the issue here. That proc is all copypaste, you should maybe refactor this code to use the material container.

**Comment by Profakos** [comment]
Fixed the parameter.

**Comment by Profakos** [comment]
Ah, I see the issue was the copypaste.

I tried to make this more dynamic before i gave up, but i guess i should have another go at it.

**Comment by Profakos** [comment]
I was thinking I would use associated list like the ore redemption machine, or an infinite material holder. Because if I gave it a storage limit, and it failed to insert the ore into the machine, and had it stay behind the conveyor belt, it would stay there forever as HasProximity is only triggered during a Crossed.

File: code/modules/mining/machine_processing.dm [hunk 2]
```diff
@@ -206,6 +219,7 @@
 	var/ore_clown = 0;
 	var/ore_adamantine = 0;
 	var/ore_titanium = 0;
+	var/ore_bluespace = 0
```
Comment (Iamgoofball) [inline code/modules/mining/machine_processing.dm]: :-1:

This code is fucking atrocious please refactor completely

**Comment by Iamgoofball** [comment]
datum recipes

**Comment by Profakos** [comment]
Sounds good. Perhaps, make them design datums even (which would pave the way for allowing science to upgrade the ORM with alien alloy smelting once the freeze is over).

**Comment by Profakos** [comment]
Figured what to do with overfilling: it autosmelts some ore to make space.

**Comment by Iamgoofball** [comment]
Just remove the max cap

**Comment by Profakos** [comment]
Converted the whole thing to materials. I had to add an extra optional parameter for eject sheets, called target. Pass through anything (be it machine, turf, etc) and it will try to force move the ejected sheets there. If no target is passed, the sheets will just sit there.

Getting prepared for Phase 2: plasteel and plastitanium smelting refactor. And to make the menu look nicer.

**Comment by xxalpha** [comment]
Now that is a clean up.

**Comment by Profakos** [comment]
Added designs for alloys. 

I suppose this is ready, unless people want me to also remake, and include, slag making.

**Comment by Profakos** [comment]
Can someonoe remove the Work In Progress tag please? I thought the server bot would automatically remove it the same way it was added automatically.

**Comment by AnturK** [comment]
This might be asking a lot but could you make ORM use these alloy datums too. Right now ORM alloy code is a mess.

**Comment by Profakos** [comment]
I can try when I am back from vacation, but that one might also warrant a full clean up, as it does't use materials either

File: code/modules/mining/machine_processing.dm [hunk 3]
```diff
@@ -196,117 +62,83 @@
...
+	var/list/categories = list("Alloys")
+
+/obj/machinery/mineral/processing_unit/Initialize()
+	. = ..()
+	proximity_monitor = new(src, 1)
+	materials = new(src, list(MAT_METAL, MAT_GLASS, MAT_SILVER, MAT_GOLD, MAT_DIAMOND, MAT_PLASMA, MAT_URANIUM, MAT_BANANIUM, MAT_TITANIUM, MAT_BLUESPACE),PROCESSING_UNIT_CAPACITY)
+	files = new /datum/research/smelter(src)
+
+/obj/machinery/mineral/processing_unit/Destroy()
+	qdel(materials)
```
Comment (optimumtact) [inline code/modules/mining/machine_processing.dm]: you should null the reference to the console here

Comment (optimumtact) [inline code/modules/mining/machine_processing.dm]: in fact really that relationship should be two way, so the console can also null that reference if it gets deleted.

Comment (optimumtact) [inline code/modules/mining/machine_processing.dm]: because otherwise this CONSOLe could refer to a machine that no longer exists.

**Comment by lzimann** [comment]
Honestly, we should just remove this smelter and use only the ORM already.

**Comment by xxalpha** [comment]
This one was more fun, to be honest. ORM is gaymey shit with the pointless points thing.

**Comment by lzimann** [comment]
Regardless if it's this or ORM, but right now we have two machines that do the same thing.

**Comment by Profakos** [comment]
That is true, especially since this cleanup removes the "guess the alloy combo or get slag" thing.

**Comment by Profakos** [comment]
Made the requested reference cleanup, also made the max size of the materials in the smelter equal to the INFINITY define for now.
I will start working on the ORM soon. It and the smelter might need a shared parent.

**Comment by optimumtact** [comment]
So is this ready or not?

**Comment by Profakos** [comment]
I have decided to put the ORM rewrite into a different PR. This is ready for now.

**Pull request merged**